File: src/commands/restore.rs
use serde::Serialize;
use serde_json::Value;

use crate::api::{ApiClient, ApiError};
use crate::output::{
    OutputFormat, extract_pagination, format_archivable_type, format_option, print_json,
    print_key_value, print_message, print_pagination, print_table,
};

#[derive(Debug, Serialize)]
struct ListRestoresQuery {
    #[serde(skip_serializing_if = "Option::is_none")]
    r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    site_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    environment_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    backup_id: Option<String>,
    page: u32,
    per_page: u32,
}

#[derive(Debug, Serialize)]
struct CreateRestoreRequest {
    backup_id: String,
    scope: String,
}

pub fn list(
    client: &ApiClient,
    site_id: Option<String>,
    environment_id: Option<String>,
    restore_type: Option<String>,
    backup_id: Option<String>,
    page: u32,
    per_page: u32,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let query = ListRestoresQuery {
        r#type: restore_type,
        site_id,
        environment_id,
        backup_id,
        page,
        per_page,
    };
    let response: Value = client.get_with_query("/api/v1/vector/restores", &query)?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let restores = response["data"]
        .as_array()
        .ok_or_else(|| ApiError::Other("Invalid response format".to_string()))?;

    if restores.is_empty() {
        print_message("No restores found.");
        return Ok(());
    }

    let rows: Vec<Vec<String>> = restores
        .iter()
        .map(|r| {
            vec![
                r["id"].as_str().unwrap_or("-").to_string(),
                r["archivable_type"]
                    .as_str()
                    .map(format_archivable_type)
                    .unwrap_or_else(|| "-".to_string()),
                r["vector_backup_id"].as_str().unwrap_or("-").to_string(),
                r["scope"].as_str().unwrap_or("-").to_string(),
                r["status"].as_str().unwrap_or("-").to_string(),
                format_option(&r["created_at"].as_str().map(String::from)),
            ]
        })
        .collect();

    print_table(
        vec!["ID", "Model", "Backup ID", "Scope", "Status", "Created"],
        rows,
    );

    if let Some((current, last, total)) = extract_pagination(&response) {
        print_pagination(current, last, total);
    }

    Ok(())
}

pub fn show(
    client: &ApiClient,
    restore_id: &str,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let response: Value =
        client.get(&format!("/api/v1/vector/restores/{}", restore_id))?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let restore = &response["data"];

    print_key_value(vec![
        ("ID", restore["id"].as_str().unwrap_or("-").to_string()),
        (
            "Model",
            restore["archivable_type"]
                .as_str()
                .map(format_archivable_type)
                .unwrap_or_else(|| "-".to_string()),
        ),
        (
            "Model ID",
            restore["archivable_id"]
                .as_str()
                .unwrap_or("-")
                .to_string(),
        ),
        (
            "Backup ID",
            restore["vector_backup_id"]
                .as_str()
                .unwrap_or("-")
                .to_string(),
        ),
        (
            "Scope",
            restore["scope"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Trigger",
            restore["trigger"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Status",
            restore["status"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Error Message",
            format_option(&restore["error_message"].as_str().map(String::from)),
        ),
        (
            "Duration (ms)",
            format_option(&restore["duration_ms"].as_u64().map(|d| d.to_string())),
        ),
        (
            "Started At",
            format_option(&restore["started_at"].as_str().map(String::from)),
        ),
        (
            "Completed At",
            format_option(&restore["completed_at"].as_str().map(String::from)),
        ),
        (
            "Created At",
            format_option(&restore["created_at"].as_str().map(String::from)),
        ),
        (
            "Updated At",
            format_option(&restore["updated_at"].as_str().map(String::from)),
        ),
    ]);

    Ok(())
}

pub fn create(
    client: &ApiClient,
    backup_id: &str,
    scope: &str,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let body = CreateRestoreRequest {
        backup_id: backup_id.to_string(),
        scope: scope.to_string(),
    };

    let response: Value = client.post("/api/v1/vector/restores", &body)?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let restore = &response["data"];
    let restore_id = restore["id"].as_str().unwrap_or("-");

    print_message(&format!(
        "Restore initiated. Use `vector restore show {}` to check progress.",
        restore_id
    ));

    print_key_value(vec![
        ("ID", restore_id.to_string()),
        (
            "Backup ID",
            restore["vector_backup_id"]
                .as_str()
                .unwrap_or("-")
                .to_string(),
        ),
        (
            "Scope",
            restore["scope"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Status",
            restore["status"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Created At",
            format_option(&restore["created_at"].as_str().map(String::from)),
        ),
    ]);

    Ok(())
}

File: internal/commands/restore.go
package commands

import (
	"encoding/json"
	"fmt"
	"io"

	"github.com/spf13/cobra"

	"github.com/built-fast/vector-cli/internal/output"
)

const restoresBasePath = "/api/v1/vector/restores"

// NewRestoreCmd creates the restore command group.
func NewRestoreCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "restore",
		Short: "Manage restores",
		Long:  "Manage restores to recover site data from backups.",
	}

	cmd.AddCommand(newRestoreListCmd())
	cmd.AddCommand(newRestoreShowCmd())
	cmd.AddCommand(newRestoreCreateCmd())

	return cmd
}

func newRestoreListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List restores",
		Long:  "Retrieve a paginated list of restores, optionally filtered by type, site, environment, or backup.",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			page, perPage := getPagination(cmd)
			query := buildPaginationQuery(page, perPage)

			if cmd.Flags().Changed("site-id") {
				v, _ := cmd.Flags().GetString("site-id")
				if v != "" {
					query.Set("site_id", v)
				}
			}
			if cmd.Flags().Changed("environment-id") {
				v, _ := cmd.Flags().GetString("environment-id")
				if v != "" {
					query.Set("environment_id", v)
				}
			}
			if cmd.Flags().Changed("type") {
				v, _ := cmd.Flags().GetString("type")
				if v != "" {
					query.Set("type", v)
				}
			}
			if cmd.Flags().Changed("backup-id") {
				v, _ := cmd.Flags().GetString("backup-id")
				if v != "" {
					query.Set("backup_id", v)
				}
			}

			resp, err := app.Client.Get(cmd.Context(), restoresBasePath, query)
			if err != nil {
				return fmt.Errorf("failed to list restores: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to list restores: %w", err)
			}

			if app.Format == output.JSON {
				data, err := parseResponseData(body)
				if err != nil {
					return fmt.Errorf("failed to list restores: %w", err)
				}
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			data, meta, err := parseResponseWithMeta(body)
			if err != nil {
				return fmt.Errorf("failed to list restores: %w", err)
			}

			var items []map[string]any
			if err := json.Unmarshal(data, &items); err != nil {
				return fmt.Errorf("failed to list restores: %w", err)
			}

			headers := []string{"ID", "MODEL", "BACKUP ID", "SCOPE", "STATUS", "CREATED"}
			var rows [][]string
			for _, item := range items {
				rows = append(rows, []string{
					getString(item, "id"),
					formatArchivableType(getString(item, "archivable_type")),
					getString(item, "vector_backup_id"),
					getString(item, "scope"),
					getString(item, "status"),
					getString(item, "created_at"),
				})
			}

			output.PrintTable(cmd.OutOrStdout(), headers, rows)
			printPaginationIfNeeded(cmd.OutOrStdout(), meta)
			return nil
		},
	}
	addPaginationFlags(cmd)
	cmd.Flags().String("site-id", "", "Filter by site ID")
	cmd.Flags().String("environment-id", "", "Filter by environment ID")
	cmd.Flags().String("type", "", "Filter by type (site/environment)")
	cmd.Flags().String("backup-id", "", "Filter by backup ID")
	return cmd
}

func newRestoreShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show a restore",
		Long:  "Display details of a specific restore.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			resp, err := app.Client.Get(cmd.Context(), restoresBasePath+"/"+args[0], nil)
			if err != nil {
				return fmt.Errorf("failed to get restore: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to get restore: %w", err)
			}

			data, err := parseResponseData(body)
			if err != nil {
				return fmt.Errorf("failed to get restore: %w", err)
			}

			if app.Format == output.JSON {
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			var item map[string]any
			if err := json.Unmarshal(data, &item); err != nil {
				return fmt.Errorf("failed to get restore: %w", err)
			}

			output.PrintKeyValue(cmd.OutOrStdout(), []output.KeyValue{
				{Key: "ID", Value: getString(item, "id")},
				{Key: "Model", Value: formatArchivableType(getString(item, "archivable_type"))},
				{Key: "Model ID", Value: getString(item, "archivable_id")},
				{Key: "Backup ID", Value: getString(item, "vector_backup_id")},
				{Key: "Scope", Value: getString(item, "scope")},
				{Key: "Trigger", Value: getString(item, "trigger")},
				{Key: "Status", Value: getString(item, "status")},
				{Key: "Error Message", Value: formatString(getString(item, "error_message"))},
				{Key: "Duration", Value: formatFloat(getFloat(item, "duration_ms"))},
				{Key: "Started At", Value: formatString(getString(item, "started_at"))},
				{Key: "Completed At", Value: formatString(getString(item, "completed_at"))},
				{Key: "Created At", Value: getString(item, "created_at")},
				{Key: "Updated At", Value: getString(item, "updated_at")},
			})
			return nil
		},
	}
}

func newRestoreCreateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "create <backup-id>",
		Short: "Create a restore",
		Long:  "Create a new restore from a backup.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			reqBody := map[string]any{
				"vector_backup_id": args[0],
			}

			dropTables, _ := cmd.Flags().GetBool("drop-tables")
			if dropTables {
				reqBody["drop_tables"] = true
			}

			disableForeignKeys, _ := cmd.Flags().GetBool("disable-foreign-keys")
			if disableForeignKeys {
				reqBody["disable_foreign_keys"] = true
			}

			srFrom, _ := cmd.Flags().GetString("search-replace-from")
			srTo, _ := cmd.Flags().GetString("search-replace-to")
			if srFrom != "" && srTo != "" {
				reqBody["search_replace"] = []map[string]string{
					{"from": srFrom, "to": srTo},
				}
			}

			resp, err := app.Client.Post(cmd.Context(), restoresBasePath, reqBody)
			if err != nil {
				return fmt.Errorf("failed to create restore: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to create restore: %w", err)
			}

			data, err := parseResponseData(body)
			if err != nil {
				return fmt.Errorf("failed to create restore: %w", err)
			}

			if app.Format == output.JSON {
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			var item map[string]any
			if err := json.Unmarshal(data, &item); err != nil {
				return fmt.Errorf("failed to create restore: %w", err)
			}

			w := cmd.OutOrStdout()
			output.PrintMessage(w, fmt.Sprintf("Restore initiated. Use 'vector restore show %s' to check progress.", getString(item, "id")))
			output.PrintMessage(w, "")

			output.PrintKeyValue(w, []output.KeyValue{
				{Key: "ID", Value: getString(item, "id")},
				{Key: "Model", Value: formatArchivableType(getString(item, "archivable_type"))},
				{Key: "Model ID", Value: getString(item, "archivable_id")},
				{Key: "Backup ID", Value: getString(item, "vector_backup_id")},
				{Key: "Scope", Value: getString(item, "scope")},
				{Key: "Trigger", Value: getString(item, "trigger")},
				{Key: "Status", Value: getString(item, "status")},
				{Key: "Error Message", Value: formatString(getString(item, "error_message"))},
				{Key: "Duration", Value: formatFloat(getFloat(item, "duration_ms"))},
				{Key: "Started At", Value: formatString(getString(item, "started_at"))},
				{Key: "Completed At", Value: formatString(getString(item, "completed_at"))},
				{Key: "Created At", Value: getString(item, "created_at")},
				{Key: "Updated At", Value: getString(item, "updated_at")},
			})
			return nil
		},
	}

	cmd.Flags().Bool("drop-tables", false, "Drop existing tables before restore")
	cmd.Flags().Bool("disable-foreign-keys", false, "Disable foreign key checks during restore")
	cmd.Flags().String("search-replace-from", "", "URL to search for (used with --search-replace-to)")
	cmd.Flags().String("search-replace-to", "", "URL to replace with (used with --search-replace-from)")

	return cmd
}
